trie/goroutinesManager_test.go

```diff
@@ -0,0 +1,177 @@
+		assert.NotNil(t, manager)
+		assert.Nil(t, err)
+		assert.True(t, manager.(*goroutinesManager).canProcess)
+	})
+}
+
+func TestGoroutinesManager_ShouldContinueProcessing(t *testing.T) {
+	t.Parallel()
+
+	t.Run("should return false if chan is closed", func(t *testing.T) {
+		t.Parallel()
+
+		closeChan := make(chan struct{})
+		manager, _ := NewGoroutinesManager(&mock.ThrottlerStub{}, errChan.NewErrChanWrapper(), closeChan)
+
+		close(closeChan)
+		assert.False(t, manager.ShouldContinueProcessing())
+	})
+	t.Run("should return true if canProcess", func(t *testing.T) {
+		t.Parallel()
+
+		closeChan := make(chan struct{})
+		manager, _ := NewGoroutinesManager(&mock.ThrottlerStub{}, errChan.NewErrChanWrapper(), closeChan)
+
+		assert.True(t, manager.ShouldContinueProcessing())
+	})
+	t.Run("should return false if !canProcess", func(t *testing.T) {
+		t.Parallel()
+
+		manager := &goroutinesManager{
+			chanClose:  make(chan struct{}),
+			canProcess: false,
+		}
+
+		assert.False(t, manager.ShouldContinueProcessing())
+	})
+}
+
+func TestGoroutinesManager_CanStartGoRoutine(t *testing.T) {
+	t.Parallel()
+
+	t.Run("should return false if !throttler.CanProcess", func(t *testing.T) {
+		t.Parallel()
+
+		throttler := &mock.ThrottlerStub{
+			CanProcessCalled: func() bool {
+				return false
+			},
+		}
+		manager, _ := NewGoroutinesManager(throttler, errChan.NewErrChanWrapper(), make(chan struct{}))
+
+		assert.False(t, manager.CanStartGoRoutine())
+	})
+	t.Run("should return true if throttler.CanProcess", func(t *testing.T) {
+		t.Parallel()
+
+		startProcessingCalled := false
+		throttler := &mock.ThrottlerStub{
+			CanProcessCalled: func() bool {
+				return true
+			},
+			StartProcessingCalled: func() {
+				startProcessingCalled = true
+			},
+		}
+		manager, _ := NewGoroutinesManager(throttler, errChan.NewErrChanWrapper(), make(chan struct{}))
+
+		assert.True(t, manager.CanStartGoRoutine())
+		assert.True(t, startProcessingCalled)
+	})
+}
+
+func TestGoroutinesManager_EndGoRoutineProcessing(t *testing.T) {
+	t.Parallel()
+
+	endProcessingCalled := false
+	throttler := &mock.ThrottlerStub{
+		EndProcessingCalled: func() {
+			endProcessingCalled = true
+		},
+	}
+	manager, _ := NewGoroutinesManager(throttler, errChan.NewErrChanWrapper(), make(chan struct{}))
+	manager.EndGoRoutineProcessing()
+	assert.True(t, endProcessingCalled)
+}
+
+func TestGoroutinesManager_SetError(t *testing.T) {
+	t.Parallel()
+
+	t.Run("should set error", func(t *testing.T) {
+		expectedErr := errors.New("error")
+		errCh := errChan.NewErrChanWrapper()
+		manager, _ := NewGoroutinesManager(&mock.ThrottlerStub{}, errCh, make(chan struct{}))
+
+		err := errCh.ReadFromChanNonBlocking()
+		assert.Nil(t, err)
+
+		manager.SetError(expectedErr)
+
+		err = errCh.ReadFromChanNonBlocking()
+		assert.Equal(t, expectedErr, err)
+		assert.False(t, manager.(*goroutinesManager).canProcess)
```

Comment:
maybe better expose `canProcess` in export_test.go to avoid casting here?

Reply:
Actually, the problem was that the constructor was returning an interface, not the actual type. Changed the constructor.